src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+
+  actions.forEach((action) => {
+    const previousState =
+      states.length > 0 ? states[states.length - 1] : { ...state };
+    let newState = { ...previousState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
+        break;
+
+      default:
+        break;
+    }
+
+    states.push(newState);
```

Comment:
[DONT REPEAT YOURSELF] - Pushing the `newState` to the `states` array is done after each case in the `switch` statement. This action should be performed just once afterwards, outside of the `switch` block, after the current action is processed to avoid repetition.